pkg/plugins/manager/pipeline/stages/discovery/discovery.go

```diff
@@ -4,69 +4,54 @@ import (
 	"context"
 
 	"github.com/grafana/grafana/pkg/plugins"
-	"github.com/grafana/grafana/pkg/plugins/config"
 	"github.com/grafana/grafana/pkg/plugins/log"
 	"github.com/grafana/grafana/pkg/plugins/manager/loader/finder"
 	"github.com/grafana/grafana/pkg/plugins/manager/registry"
 )
 
 type Discoverer interface {
-	Discover(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error)
+	Run(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error)
 }
 
-type FindFunc func(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error)
-
-type FindFilterFunc func(ctx context.Context, bundles []*plugins.FoundBundle) ([]*plugins.FoundBundle, error)
+type FindStep func(ctx context.Context, prev []*plugins.FoundBundle) ([]*plugins.FoundBundle, error)
 
 type Discovery struct {
-	findStep       FindFunc
-	findFilterStep FindFilterFunc
-	log            log.Logger
-}
-
-type Opts struct {
-	FindFunc       FindFunc
-	FindFilterFunc FindFilterFunc
+	// Services
+	pluginFinder   finder.Finder
+	pluginRegistry registry.Service
+	// Assets
+	src plugins.PluginSource
+	// Misc
+	log log.Logger
 }
 
-func New(cfg *config.Cfg, opts Opts) *Discovery {
-	if opts.FindFunc == nil {
-		opts.FindFunc = DefaultFindFunc(cfg)
-	}
-
-	if opts.FindFilterFunc == nil {
-		opts.FindFilterFunc = DefaultFindFilterFunc
-	}
-
-	return &Discovery{
-		findStep:       opts.FindFunc,
-		findFilterStep: opts.FindFilterFunc,
+func New(pluginFinder finder.Finder, pluginRegistry registry.Service) *Discovery {
```

Comment:
If I want to have different step implementations, does this mean I have to implement a new `Discoverer` with this approach? The `Discover` struct has now changed to a Grafana specific implementation, and now there's no place to customize different different steps - or have I missed something? 🤔

Reply:
I am not sure I follow so to add more color: If a new external input to the stage is needed, then the `Discoverer` interface needs to change to accept that new field (just like a new service dependency).

I am not sure I understand the limitation for customization you mention. With this, steps has access to two things: The stage inputs (read-only) through the stage struct (`Discover` in this case) and the asset being processed (`[]*plugins.FoundBundle`). With that the step should have everything to produce a new "version" of the asset.

I believe that the only change in concept here is that with your approach, the customization is in the steps while with this approach the customization is in the constructor. You save the need to think about the interface for new steps but you lose traceability for which step require which input.

Reply:
Maybe an example would help here. So we have the `Bootstrap` stage:

```go
type Bootstrap struct {
	constructStep ConstructFunc
	decorateSteps []DecorateFunc
	log           log.Logger
}

type Opts struct {
	ConstructFunc ConstructFunc
	DecorateFuncs []DecorateFunc
}

func New(cfg *config.Cfg, opts Opts) *Bootstrap {
	if opts.ConstructFunc == nil {
		opts.ConstructFunc = DefaultConstructFunc(signature.DefaultCalculator(), assetpath.DefaultService(cfg))
	}

	if len(opts.DecorateFuncs) == 0 {
		opts.DecorateFuncs = DefaultDecorateFuncs
	}

	return &Bootstrap{
		constructStep: opts.ConstructFunc,
		decorateSteps: opts.DecorateFuncs,
		log:           log.New("plugins.discovery"),
	}
}

func (b *Bootstrap) Bootstrap(ctx context.Context, src plugins.PluginSource, found []*plugins.FoundBundle) ([]*plugins.Plugin, error) {
	ps, err := b.constructStep(ctx, src, found)
	if err != nil {
		return nil, err
	}

	for _, p := range ps {
		for _, decorator := range b.decorateSteps {
			p, err = decorator(ctx, p)
			if err != nil {
				return nil, err
			}
		}
	}

	return ps, nil
}
```

This means that in `pkg/services/pluginsintegration/pipeline`, Grafana can have its own implementation:

```go
func ProvideDiscoveryStage(cfg *config.Cfg, pluginFinder finder.Finder, pluginRegistry registry.Service) *discovery.Discovery {
	return discovery.New(cfg, discovery.Opts{
		FindFunc: func(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error) {
			return pluginFinder.Find(ctx, src)
		},
		FindFilterFunc: func(ctx context.Context, bundles []*plugins.FoundBundle) ([]*plugins.FoundBundle, error) {
			return discovery.NewDuplicatePluginFilterStep(pluginRegistry).Filter(ctx, bundles)
		},
	})
}
```

It also means that the runner could do it's own thing:

```go
d := discovery.New(cfg, discovery.Opts{
	FindFunc: func(ctx context.Context, src plugins.PluginSource) ([]*plugins.FoundBundle, error) {
		  var res []*plugins.FoundBundle
		  for _, u := range src.PluginURIs(ctx) {
			  p := s.fetchPlugin(ctx, u)
			  res = append(res, p)
		  }
		  return res, nil
		},
		FindFilterFunc: func(_ context.Context, bundles []*plugins.FoundBundle) ([]*plugins.FoundBundle, error) {
			return bundles // no filter
		},
	})
}
```

It's just unclear to me how one would achieve the same thing with this approach, where we are able to have granular control over different steps of the pipeline, depending on the context we're loading plugins from.

Reply:
aah, for that yes, we would need to have a `Discoverer` for Grafana and another for the runner. Each one, potentially with a different number of steps. If sharing steps is something desirable (I guess so), we would need to expose a `FindFunc`/`FindStep` that both `Discoverer` could use. That would still work with this model but we would be losing the general (hidden) definition of steps, making it redundant :shrug:. It makes less sense in that case.

Anyway, feel free to disregard the proposal, that's a fair point.